// gatsby-remark-typescript/index.js
"use strict";
var visit = require("unist-util-visit");
var prettier_1 = require("prettier");
var core_1 = require("@babel/core");
function insertAt(array, index, item) {
    return array.slice(0, index).concat([item], array.slice(index));
}
function isReactComponent(node, tag) {
    return node && node.type === 'jsx' && node.value === tag;
}
var PRESERVE_COMMENT = '// preserve-line';
var PRESERVE_PATTERN = new RegExp("\\s" + PRESERVE_COMMENT + "$");
function removePreserveComment(line) {
    return line.replace(PRESERVE_PATTERN, '');
}
module.exports = function plugin(_a, _b) {
    var markdownAST = _a.markdownAST;
    var _c = _b === void 0 ? {} : _b, wrapperComponent = _c.wrapperComponent, _d = _c.prettierOptions, prettierOptions = _d === void 0 ? { parser: 'babel' } : _d;
    function visitor(node, index, parent) {
        if (/^tsx?/.test(node.lang)) {
            var prevNode = parent.children[index - 1];
            var nextNode = parent.children[index + 1];
            if (wrapperComponent) {
                var isWrapped = isReactComponent(prevNode, "<" + wrapperComponent + ">") &&
                    isReactComponent(nextNode, "</" + wrapperComponent + ">");
                if (!isWrapped) {
                    return;
                }
            }
            try {
                var lines = node.value.split('\n');
                var code = core_1.transformSync(lines
                    .map(function (line) {
                    return PRESERVE_PATTERN.test(line) ? '// ' + line : line;
                })
                    .join('\n'), {
                    filename: "file." + node.lang,
                    retainLines: true,
                    presets: ['@babel/typescript']
                }).code;
                if (code.trim()) {
                    parent.children = insertAt(parent.children, index + 1, {
                        type: 'code',
                        lang: node.lang.replace(/^ts|(?<=\.)ts/g, 'js'),
                        value: prettier_1.format(code
                            .split('\n')
                            .map(function (line) {
                            return PRESERVE_PATTERN.test(line)
                                ? removePreserveComment(line.replace(/^\/\/\s/, ''))
                                : line;
                        })
                            .join('\n'), prettierOptions).trim()
                    });
                }
                node.value = lines.map(removePreserveComment).join('\n');
            }
            catch (error) {
                console.error(error.message);
            }
        }
    }
    visit(markdownAST, 'code', visitor);
    return markdownAST;
};
